File: src/main/java/features/repositoryaccessors/RepoFileSaver.java
package features.repositoryaccessors;

import core.algorithm.Algorithm;
import core.data.Data;
import core.data.FileData;
import core.repository.RepositoryCommanderException;
import features.repositoryaccessors.data.RepoRequest;

/**
 * {@link Algorithm} that saves file, presented in {@link FileData} using provided {@link RepoRequest}
 */
public class RepoFileSaver implements Algorithm<RepoRequest, Data> {

    @Override
    public Data execute(RepoRequest input) {

        if(!(input.bundle instanceof FileData))
            throw new RuntimeException("invalid request: FileData parse error");

        FileData data = (FileData)input.bundle;

        try {
            input.getRepositoryCommander().saveFile(data.fileBytes, input.getLink(), data.name + data.extension);
        } catch (Exception e) {
            throw new AlgorithmExecutionException("Unable to save file", e);
        }

        return Data.empty;
    }
}

File: src/main/java/features/repositoryaccessors/RepoFileSaver.kt
package features.repositoryaccessors

import core.algorithm.Algorithm
import core.data.Data
import core.data.FileData
import features.repositoryaccessors.data.RepoRequest

/**
 * [Algorithm] that saves file, presented in [FileData] using provided [RepoRequest]
 */
class RepoFileSaver : Algorithm<RepoRequest, Data> {

    override fun execute(input: RepoRequest): Data {

        if (input.bundle !is FileData)
            throw RuntimeException("invalid request: FileData parse error")

        val data = input.bundle as FileData?

        try {
            input.repositoryCommander.saveFile(data!!.fileBytes, input.link, data.name + data.extension)
        } catch (e: Exception) {
            throw AlgorithmExecutionException("Unable to save file", e)
        }

        return Data.empty
    }
}
